### ISSUE TITLE
ScErrDataBindingForJsonException for PageSize$ when PageSize is tried to be created before object is created or initialized

### ISSUE THREAD

**Issue by Razkan** [Starcounter/Starcounter.Uniform#18]
Starcounter version: `<2.4.990.175>`.
Starcounter.Uniform version: `<2.1.1>`.

### Issue type

- [x] Bug

### Issue description

Getting this issue since `PageSize$` tries to bind before an `DataTableBuilder` has been built for the entity. which happens when the following handler is called: 

```cs
// BlendingProvider.Website.Api.ContentHandlers` at line 243
result = Self.GET("/BlendingProvider/MasterWrapper?" +
                      $"request={requestKey}&" +
                      $"response={responseKey}");
```

The application code is next to identical to the People Apps PersonsPage line 14 to 25, with the exception that `Init` is run at `OnData`, instead of line 47 in MainHandlers.

**Note**: None of my table code is executed before the exception is thrown, this happens before an object has had the chance to be created or initialized in the app.

### Exception stack trace

```
System.Exception: ScErrDataBindingForJsonException (SCERR14012): An exception occurred when a databinding for a property in json was used. Property 'PageSize$' in '<OurApp>'.
Version: 2.4.990.175.
Help page: https://docs.starcounter.io/?q=SCERR14012.
   at lambda_method(Closure , Json )
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.GetValue[T](Property`1 template, Json parent, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 580
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeLong(Json json, Template template, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 172
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeObject(Json json, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 241
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 225
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeObject(Json json, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 241
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 225
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeObject(Json json, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 241
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 225
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeObject(Json json, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 241
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 225
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.SerializeObject(Json json, JsonWriter writer, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 241
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 225
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.<>c__DisplayClass10_1.<Serialize>b__0() in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 79
   at Starcounter.Internal.TransactionManager.Scope(TransactionHandle handle, Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\TransactionManager.cs:line 557
   at Starcounter.Transaction.Scope(Action action) in C:\tc_work\sc-19526\Level1\src\Starcounter\Transaction.cs:line 233
   at Starcounter.XSON.Internal.XSONInternals.RunWithinApplication(String appName, Action action) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Internal\XSONInternals.cs:line 41
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.Serialize(Json json, Template template, TextWriter textWriter, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 82
   at Starcounter.XSON.Serializer.NewtonSoftSerializer.Serialize(Json json, Template template, JsonSerializerSettings settings) in c:\Git\Starcounter.XSON\src\Starcounter.XSON\Serializer\NewtonSoftSerializer.cs:line 61
   at BlendingProvider.Website.ViewModels.ViewKeeper.WrapResponse(Response response) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\ViewModels\ViewKeeper.json.cs:line 62
   at BlendingProvider.Website.Api.ContentHandlers.UpdateTemplateSections(String requestUri, Response response, LayoutPage content, WebUrl url) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\Api\ContentHandlers.cs:line 305
   at BlendingProvider.Website.Api.ContentHandlers.<Register>b__14_2(String requestUri, String responseKey, String webUrlId, Boolean isOriginalResponse) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\Api\ContentHandlers.cs:line 80
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\tc_work\sc-19526\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 203
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\tc_work\sc-19526\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 95
   at Starcounter.Rest.UriManagedHandlersCodegen.RunUriMatcherAndCallHandler(String methodSpaceUriSpace, String methodSpaceUriSpaceLower, Request req, UInt16 portNumber, Response& resp) in C:\tc_work\sc-19526\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 1274
   at Starcounter.Self.DoSelfCall(UInt16 portNumber, String method, String relativeUri, Dictionary`2 headersDictionary, String body, Byte[] bodyBytes, HandlerOptions handlerOptions, Request req) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Rest\Self.cs:line 481
   at Starcounter.Self.GET(String uri) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Rest\Self.cs:line 102
   at BlendingProvider.Website.Api.ContentHandlers.WrapResponse(String requestUri, Dictionary`2 headersDictionary, Response originalResponse) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\Api\ContentHandlers.cs:line 157
   at BlendingProvider.Website.Api.ContentHandlers.<Register>b__14_5(String requestKey, String responseKey) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\Api\ContentHandlers.cs:line 134
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\tc_work\sc-19526\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 203
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\tc_work\sc-19526\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 95
   at Starcounter.Rest.UriManagedHandlersCodegen.RunUriMatcherAndCallHandler(String methodSpaceUriSpace, String methodSpaceUriSpaceLower, Request req, UInt16 portNumber, Response& resp) in C:\tc_work\sc-19526\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 1274
   at Starcounter.Self.DoSelfCall(UInt16 portNumber, String method, String relativeUri, Dictionary`2 headersDictionary, String body, Byte[] bodyBytes, HandlerOptions handlerOptions, Request req) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Rest\Self.cs:line 481
   at Starcounter.Self.GET(String uri) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Rest\Self.cs:line 102
   at BlendingProvider.Website.Api.ContentHandlers.<RegisterFilter>b__17_1(Request request, Response response) in C:\Colabers\Blending.Mapper\Blending\src\BlendingProvider\Website\Api\ContentHandlers.cs:line 243
   at Starcounter.Handle.RunResponseFilters(Request req, Response resp) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Handle.cs:line 612
   at Starcounter.Request.SendResponseInternal(Response resp, Byte[] serializationBuf) in C:\tc_work\sc-19526\Level1\src\Starcounter.Internal\Http\Request.cs:line 1123
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\tc_work\sc-19526\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 402
HResult=-2146233088
HelpLink=https://docs.starcounter.io/?q=SCERR14012
---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Starcounter.Uniform.ViewModels.UniDataTable.PaginationViewModel.get_PageSize() in C:\Users\joozek\dev\starcounter.uniform\Starcounter.Uniform\ViewModels\UniDataTable.json.cs:line 108
   at lambda_method(Closure , Json )
HResult=-2147467261

More information about this error may be available in the server error log.
```

**Comment by joozek78** [comment]
Thank you for the report. What exact revision of Blending did you use?

**Comment by Razkan** [comment]
@joozek78 `commit 1956ad4719ae6b0fc1c14dfde73ca2d487208a41 (HEAD, tag: 6.5.0)`

**Comment by Razkan** [comment]
@joozek78 Is something happening here? This issue blocks one of our apps.

**Comment by joozek78** [comment]
I still can't take it, but I forwarded it internally. Someone will answer here soon
CC @mmnosek

**Comment by stefanedqvist** [comment]
Pls @warpech can you address this?

**Comment by warpech** [comment]
@Razkan does this happen when BlendingProvider is not running? Your app should still work, just not be decorated in a layout and without custom compositions.

**Comment by Razkan** [comment]
@warpech @joozek78 the pull-request https://github.com/Starcounter/Starcounter.Uniform/pull/20#issue-225769623 was enough to solve the current issue. 

I did some basic testing, and it is not enough to conclude the root cause but when I had a fresh db, the issue did not not present itself. As soon as I changed the default composition and layout to our custom layout and composition, then reloaded the app (localhost:8080/BlendingEditor/settings) the issue occured. A `PaginationViewModel` object of the `UniDataTable` was tried to be created, which caused the issue since DataProvider had not been initialized.

Perhaps that's a lead if you wish to continue investigate. Contact me on slack if you want a copy of the custom layout.
